File: api/py/tract/transform.py
import json
from abc import ABC, abstractmethod
from typing import Dict, Optional, Union

from .tensor import DatumType


class TransformSpec(ABC):
    """Base class for typed transform specifications.

    Subclasses represent specific transforms with typed parameters.
    Can be passed directly to :meth:`Model.transform`.
    """

    @abstractmethod
    def to_json(self) -> str:
        """Serialize this transform spec to the JSON string the FFI layer expects."""
        ...


class ConcretizeSymbols(TransformSpec):
    """Replace symbolic dimensions with concrete integer values.

    Example::

        model.transform(ConcretizeSymbols({"B": 1}))
        # or with builder pattern:
        model.transform(ConcretizeSymbols().value("B", 1))
    """

    def __init__(self, values: Optional[Dict[str, int]] = None):
        self._values: Dict[str, int] = dict(values) if values else {}

    def value(self, symbol: str, val: int) -> "ConcretizeSymbols":
        """Set a symbol to a concrete value. Returns self for chaining."""
        self._values[symbol] = val
        return self

    def to_json(self) -> str:
        return json.dumps({"name": "concretize_symbols", "values": self._values})


class Pulse(TransformSpec):
    """Convert a model to a pulsed (streaming) model.

    Example::

        model.transform(Pulse("5", symbol="B"))
        # or with builder pattern:
        model.transform(Pulse("5").symbol("B"))
    """

    def __init__(self, pulse: Union[str, int], *, symbol: Optional[str] = None):
        self._pulse = str(pulse)
        self._symbol = symbol

    def symbol(self, symbol: str) -> "Pulse":
        """Set the symbol to pulse over. Returns self for chaining."""
        self._symbol = symbol
        return self

    def to_json(self) -> str:
        d = {"name": "pulse", "pulse": self._pulse}
        if self._symbol is not None:
            d["symbol"] = self._symbol
        return json.dumps(d)


class FloatPrecision(TransformSpec):
    """Change the float precision of a model (e.g. F32 to F16).

    Example::

        model.transform(FloatPrecision(DatumType.F32, DatumType.F16))
        # with include/exclude:
        model.transform(FloatPrecision(DatumType.F32, DatumType.F16, exclude=["layer.1"]))
    """

    _DT_NAMES = {
        DatumType.F16: "f16",
        DatumType.F32: "f32",
        DatumType.F64: "f64",
    }

    def __init__(
        self,
        from_dt: DatumType,
        to_dt: DatumType,
        *,
        include: Optional[list] = None,
        exclude: Optional[list] = None,
    ):
        if from_dt not in self._DT_NAMES:
            raise ValueError(f"from_dt must be a float DatumType, got {from_dt}")
        if to_dt not in self._DT_NAMES:
            raise ValueError(f"to_dt must be a float DatumType, got {to_dt}")
        self._from = from_dt
        self._to = to_dt
        self._include = list(include) if include else None
        self._exclude = list(exclude) if exclude else None

    def include(self, patterns: list) -> "FloatPrecision":
        """Set include patterns — only matching nodes are translated. Returns self for chaining."""
        self._include = list(patterns)
        return self

    def exclude(self, patterns: list) -> "FloatPrecision":
        """Set exclude patterns — matching nodes are excluded. Returns self for chaining."""
        self._exclude = list(patterns)
        return self

    def to_json(self) -> str:
        d = {
            "name": "float_precision",
            "from": self._DT_NAMES[self._from],
            "to": self._DT_NAMES[self._to],
        }
        if self._include is not None:
            d["include"] = self._include
        if self._exclude is not None:
            d["exclude"] = self._exclude
        return json.dumps(d)

File: api/src/transform.rs
use std::collections::HashMap;

use crate::DatumType;

/// A serialized transform specification passed to `ModelInterface::transform`.
///
/// Wraps the string representation expected by the transform registry.
/// Constructed from raw strings or typed config structs implementing [`TransformConfig`].
#[derive(Debug, Clone)]
pub struct TransformSpec(String);

impl TransformSpec {
    /// Produce the string the transform registry expects.
    pub fn to_transform_string(&self) -> String {
        self.0.clone()
    }
}

impl From<&str> for TransformSpec {
    fn from(s: &str) -> Self {
        TransformSpec(s.to_string())
    }
}

impl From<String> for TransformSpec {
    fn from(s: String) -> Self {
        TransformSpec(s)
    }
}

impl From<&String> for TransformSpec {
    fn from(s: &String) -> Self {
        TransformSpec(s.clone())
    }
}

/// Trait for typed transform configurations.
///
/// Implementors derive [`serde::Serialize`] and provide a transform [`name()`](TransformConfig::name).
/// The default [`to_transform_string()`](TransformConfig::to_transform_string) serializes the
/// struct as a JSON object and injects the `"name"` key.
pub trait TransformConfig: serde::Serialize {
    /// The transform registry name (e.g. `"pulse"`, `"float_precision"`).
    fn name(&self) -> &'static str;

    /// Produce the string the transform registry expects.
    ///
    /// The default implementation serializes `self` to a JSON object and inserts `"name"`.
    fn to_transform_string(&self) -> String {
        let mut obj: serde_json::Map<String, serde_json::Value> = serde_json::to_value(self)
            .expect("TransformConfig serialization cannot fail")
            .as_object()
            .expect("TransformConfig must serialize to a JSON object")
            .clone();
        obj.insert("name".into(), serde_json::Value::String(self.name().to_string()));
        serde_json::to_string(&obj).expect("serialization cannot fail")
    }
}

/// Implements [`TransformConfig`] and `From<$ty> for TransformSpec`.
macro_rules! transform_config {
    ($ty:ty, $name:expr) => {
        impl TransformConfig for $ty {
            fn name(&self) -> &'static str {
                $name
            }
        }

        impl From<$ty> for TransformSpec {
            fn from(config: $ty) -> Self {
                TransformSpec(config.to_transform_string())
            }
        }
    };
}

/// Typed config for the `set_symbols` transform.
///
/// Binds symbolic dimensions to concrete integers (or `TDim` expressions
/// via [`Self::expr`]).
///
/// # Example
/// ```ignore
/// model.transform(SetSymbols::new().value("B", 1).value("T", 16))?;
/// ```
#[derive(Debug, Clone, Default, serde::Serialize)]
pub struct SetSymbols {
    #[serde(serialize_with = "serialize_values")]
    values: HashMap<String, SetSymbolValue>,
}

#[derive(Debug, Clone, serde::Serialize)]
#[serde(untagged)]
enum SetSymbolValue {
    Int(i64),
    Expr(String),
}

fn serialize_values<S: serde::Serializer>(
    values: &HashMap<String, SetSymbolValue>,
    s: S,
) -> Result<S::Ok, S::Error> {
    use serde::ser::SerializeMap;
    let mut map = s.serialize_map(Some(values.len()))?;
    for (k, v) in values {
        map.serialize_entry(k, v)?;
    }
    map.end()
}

impl SetSymbols {
    pub fn new() -> Self {
        Self::default()
    }

    /// Bind a symbol to a concrete integer value.
    pub fn value(mut self, symbol: impl Into<String>, val: i64) -> Self {
        self.values.insert(symbol.into(), SetSymbolValue::Int(val));
        self
    }

    /// Bind a symbol to a `TDim` expression (e.g. `"2*S"`) parsed against
    /// the model's symbol scope at transform time.
    pub fn expr(mut self, symbol: impl Into<String>, expr: impl Into<String>) -> Self {
        self.values.insert(symbol.into(), SetSymbolValue::Expr(expr.into()));
        self
    }
}

transform_config!(SetSymbols, "set_symbols");

/// Typed config for the `pulse` transform.
///
/// Converts a model to a pulsed (streaming) model.
///
/// # Example
/// ```ignore
/// model.transform(Pulse::new("5").symbol("B"))?;
/// ```
#[derive(Debug, Clone, serde::Serialize)]
pub struct Pulse {
    pulse: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    symbol: Option<String>,
}

impl Pulse {
    /// Create a new Pulse config with the given pulse dimension.
    pub fn new(pulse: impl Into<String>) -> Self {
        Self { pulse: pulse.into(), symbol: None }
    }

    /// Set the symbol to pulse over (defaults to "S" if not set).
    pub fn symbol(mut self, symbol: impl Into<String>) -> Self {
        self.symbol = Some(symbol.into());
        self
    }
}

transform_config!(Pulse, "pulse");

/// Typed config for the `float_precision` transform.
///
/// Changes the float precision of a model (e.g. F32 to F16).
///
/// # Example
/// ```ignore
/// use tract_api::DatumType;
/// model.transform(FloatPrecision::new(DatumType::F32, DatumType::F16))?;
/// ```
#[derive(Debug, Clone, serde::Serialize)]
pub struct FloatPrecision {
    from: String,
    to: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    include: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    exclude: Option<Vec<String>>,
}

fn datum_type_to_str(dt: DatumType) -> &'static str {
    match dt {
        DatumType::F16 => "f16",
        DatumType::F32 => "f32",
        DatumType::F64 => "f64",
        _ => panic!("FloatPrecision only supports float datum types (F16, F32, F64)"),
    }
}

impl FloatPrecision {
    pub fn new(from: DatumType, to: DatumType) -> Self {
        Self {
            from: datum_type_to_str(from).to_string(),
            to: datum_type_to_str(to).to_string(),
            include: None,
            exclude: None,
        }
    }

    /// Set include patterns — only nodes matching at least one pattern are translated.
    pub fn include(mut self, patterns: Vec<String>) -> Self {
        self.include = Some(patterns);
        self
    }

    /// Set exclude patterns — matching nodes are excluded from translation.
    pub fn exclude(mut self, patterns: Vec<String>) -> Self {
        self.exclude = Some(patterns);
        self
    }
}

transform_config!(FloatPrecision, "float_precision");
